ZetaPushSwift/ZetaPushMacroService.swift

```diff
@@ -72,307 +84,265 @@ public enum ZetaPushMacroError: Error {
-                }
-                
-                self.clientHelper?.unsubscribe(sub!)
-                
-                if let result = messageDict["result"] as? NSDictionary {
-                    
-                    guard let zpMessage = U.resultType(json: result as! JSON) else {
-                        
-                        seal.reject(ZetaPushMacroError.decodingError)
-                        
-                        return
-                    }
-                    
-                    let completion = U(result: zpMessage, name: verb, requestId: requestId)
-                    
-                    seal.fulfill(completion)
-                }
-                if messageDict.object(forKey: "errors") != nil {
-                    if let errors = messageDict["errors"] as? NSArray {
-                        if errors.count > 0 {
-                            if let error = errors[0] as? NSDictionary {
-                                seal.reject(ZetaPushMacroError.genericFromDictionnary(error))
-                            } else {
-                                seal.reject(ZetaPushMacroError.unknowError)
-                            }
-                        }
-                    }
-                }
-                
+  }
+  
+  open func call<T : Glossy, U: AbstractMacroCompletion>(verb: String, parameters: T) -> Promise<U> {
+    return Promise { [weak self] seal in
+      let requestId = UUID().uuidString
+      let dict: [String: Any] = [
+        "name": verb,
+        "hardFail": false,
+        "parameters": parameters.toJSON(),
+        "requestId": requestId
+      ]
+      
+      var sub: Subscription?
+      let channelBlockMacroCall: ChannelSubscriptionBlock = { [weak self] messageDict -> Void in
+        // Check if the requestId is similar to the one sent
+        if messageDict.object(forKey: "requestId") != nil {
+          if let msgRequestId = messageDict["requestId"] as? String {
+            if msgRequestId != requestId {
+              return
             }
-            
-            sub = self.clientHelper?.subscribe(composeServiceChannel(verb), block: channelBlockMacroCall)
-            
-            self.clientHelper?.publish(composeServiceChannel("call"), message: dict)
-            
+          }
+        }
+        
+        guard let subscription = sub else { return }
+        self?.clientHelper?.unsubscribe(subscription)
+        
+        if let result = messageDict["result"] as? NSDictionary {
+          guard let json = result as? JSON, let zpMessage = U.resultType(json: json) else {
+            seal.reject(ZetaPushMacroError.decodingError)
+            return
+          }
+          
+          let completion = U(result: zpMessage, name: verb, requestId: requestId)
+          seal.fulfill(completion)
+        }
+        guard let errors = messageDict["errors"] as? [[String: Any]], errors.isEmpty else {
+          return
         }
+        if let error = errors.first as? NSDictionary {
```

Comment:
like before for errors. It is better to merge guard and if let to have unknown error management